auth_token/models.py

```diff
@@ -129,55 +171,357 @@ def delete(self):
+
+    def check_key(self, key):
+        return self.key == hash_key(key)
+
+    @property
+    def is_expired(self):
+        return self.expires_at and self.expires_at < timezone.now()
+
+
+class AuthorizationRequest(SmartModel):
+
+    authorization_token = models.ForeignKey(
+        verbose_name=_('authorization token'),
+        to=AuthorizationToken,
+        related_name='authorization_requests',
+        null=True,
+        blank=True,
+        on_delete=models.SET_NULL
+    )
+    user = models.ForeignKey(
+        verbose_name=_('user'),
+        to=django_settings.AUTH_USER_MODEL,
+        related_name='authorization_requests',
+        null=False,
+        blank=False,
+        on_delete=models.CASCADE
+    )
+    mobile_device = models.ForeignKey(
+        verbose_name=_('mobile device'),
+        to=MobileDevice,
+        related_name='authorization_requests',
+        null=True,
+        blank=True,
+        on_delete=models.CASCADE
+    )
+    one_time_password = models.OneToOneField(
+        verbose_name=_('one time password'),
+        to=OneTimePassword,
+        related_name='authorization_request',
+        null=True,
+        blank=True,
+        on_delete=models.CASCADE
+    )
+    slug = models.SlugField(
+        verbose_name=_('slug'),
+        null=True,
+        blank=True
+    )
+    title = models.CharField(
+        verbose_name=_('title'),
+        max_length=250,
+        null=False,
+        blank=False
+    )
+    description = models.TextField(
+        verbose_name=_('description'),
+        null=True,
+        blank=True
+    )
+    description_context_data = models.JSONField(
+        verbose_name=_('description context data'),
+        null=True,
+        blank=True,
+        encoder=DjangoJSONEncoder
+    )
+    result = NumEnumField(
+        verbose_name=_('result'),
+        enum=AuthorizationRequestResult,
+        null=True,
+        blank=True
+    )
+    type = NumEnumField(
+        verbose_name=_('type'),
+        enum=AuthorizationRequestType,
+        null=False,
+        blank=False
+    )
+    data = models.JSONField(
+        verbose_name=_('data'),
+        null=True,
+        blank=True,
+        encoder=DjangoJSONEncoder
+    )
+    expires_at = models.DateTimeField(
+        verbose_name=_('expires at'),
+        null=True,
+        blank=True,
+    )
+    related_objects = GenericManyToManyField()
+
+    class Meta:
+        ordering = ('-created_at',)
+        verbose_name = _('authorization request')
+        verbose_name_plural = _('authorization requests')
+
+    @property
+    def is_expired(self):
+        return self.expires_at < timezone.now()
+
+    @property
+    def state(self):
+        if not self.result and self.is_expired:
+            return AuthorizationRequestState.EXPIRED
+        elif not self.result:
+            return AuthorizationRequestState.WAITING
+        else:
+            return AuthorizationRequestState(self.result.value)
```

Comment:
Um, this is error prone, you are now basically unable to add new `AuthorizationRequestResult`, since `AuthorizationRequestState` has already value "3" taken. And no exception here would be raised. Maybe better match the enum by name, not by value?

Reply:
yes good point